src/main/java/com/google/collegeplanner/servlets/DatastoreServlet.java

```diff
@@ -0,0 +1,179 @@
+  }
+
+  public DatastoreServlet(DatastoreService datastore, ApiUtil apiUtil) {
+    super(apiUtil);
+    this.datastore = datastore;
+  }
+
+  /*
+   * Downloads all course data to datastore.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Loop through the pages.
+    // We don't know how many pages there are beforehand. We stop looping when the new page
+    // doesn't have any results.
+    int page = 1;
+    do {
+      URI uri;
+      try {
+        URIBuilder builder = new URIBuilder("https://api.umd.io/v1/courses");
+        builder.setParameter("page", Integer.toString(page));
+        uri = builder.build();
+      } catch (URISyntaxException e) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      }
+      JSONArray coursesArray = apiUtil.getJsonArray(uri);
+      if (coursesArray == null) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      } else if (coursesArray.size() == 0) {
+        // Successful. The default status code response is 200.
+        return;
+      }
+      addCourses(coursesArray, response);
+    } while (page++ < PAGE_LIMIT);
+  }
+
+  /**
+   * Creates a Course object from a JSONArray.
+   * @param coursesArray The course json from the UMD API.
+   * @param response The HttpServletResponse object.
+   */
+  private void addCourses(JSONArray coursesArray, HttpServletResponse response) throws IOException {
+    if (coursesArray == null) {
+      return;
+    }
+
+    // Loop through each page's courses.
+    for (Object jsonObject : coursesArray) {
+      JSONObject courseJson = (JSONObject) jsonObject;
+
+      // Create a Course object from the JSONObject for parsing.
+      Course course;
+      try {
+        course = new Course(courseJson);
+      } catch (Exception e) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      }
+
+      // Check if the entity already exists in datastore, and modify it if it does.
+      Query query = new Query("Course").setFilter(
+          new FilterPredicate("course_id", FilterOperator.EQUAL, course.getCourseId()));
+      PreparedQuery preparedQuery = datastore.prepare(query);
+      List<Entity> limitedResults = preparedQuery.asList(FetchOptions.Builder.withLimit(1));
+      Entity courseEntity;
+      if (limitedResults.size() == 0) {
+        // The entity doesn't exist - create a new one.
+        courseEntity = new Entity("Course");
+      } else {
+        // The entity already exists - modify it instead.
+        courseEntity = limitedResults.get(0);
+      }
+
+      URI uri;
+      try {
+        uri = new URI("https://api.umd.io/v1/courses/"
+            + URLEncoder.encode(course.getCourseId(), StandardCharsets.UTF_8.toString())
+            + "/sections");
```

Comment:
Do we have to call .../sections specifically in order to get the sections information? I am wondering if we can get that from the first api call and store it in Course?

Reply:
yes we have to call /sections. calling /course only gives us the section_ids basically, but we have to call /sections to get the full detailed info about each section